File: src/merkle/getMerkleProof.ts
// Copyright (C) 2022 Deliberative Technologies P.C.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import dcryptoMemory from "./memory";

import sha512 from "../hash/sha512";
import isUint8Array from "../utils/isUint8Array";

import dcryptoMethodsModule from "../c/build/dcryptoMethodsModule";

import { crypto_hash_sha512_BYTES } from "../utils/interfaces";

/**
 * @function
 * getMerkleProof
 *
 * @description
 * Returns the Merkle proof of an element of a tree.
 * Can be used as a receipt of a transaction etc.
 *
 * @param {(T | Uint8Array)[]} tree: The tree.
 * @param {T | Uint8Array} element: The element.
 * @param {(i: T) => Uint8Array} serializer?: Converts leaves into Uint8Array.
 *
 * @returns {Promise<Uint8Array>}: The Merkle proof.
 */
const getMerkleProof = async <T>(
  tree: (T | Uint8Array)[],
  element: T | Uint8Array,
  serializer?: (i: T) => Uint8Array,
): Promise<Uint8Array> => {
  const treeLen = tree.length;
  if (treeLen === 0) {
    throw new Error("Cannot calculate Merkle proof of element of empty tree.");
  } else if (treeLen === 1) {
    // "No point in calculating proof of a tree with single leaf.",
    return new Uint8Array(crypto_hash_sha512_BYTES + 1).fill(1);
  }

  const leavesAreUint8Arrays = isUint8Array(tree[0]);
  const elementIsUint8Array = isUint8Array(element);
  if (!serializer && (!leavesAreUint8Arrays || !elementIsUint8Array))
    throw new Error(
      "It is mandatory to provide a serializer for non-Uint8Array items",
    );

  const wasmMemory = dcryptoMemory.getMerkleProofMemory(treeLen);
  const module = await dcryptoMethodsModule({
    wasmMemory,
  });

  const ptr1 = module._malloc(treeLen * crypto_hash_sha512_BYTES);
  const leavesHashed = new Uint8Array(
    module.HEAPU8.buffer,
    ptr1,
    treeLen * crypto_hash_sha512_BYTES,
  );

  let i = 0;
  let leafIsUint8Array = false;
  let hash: Uint8Array;
  let serialized: Uint8Array;
  let leaf: T | Uint8Array;
  for (let j = 0; j < treeLen; j++) {
    leaf = tree[i];
    leafIsUint8Array = isUint8Array(leaf);
    serialized = leafIsUint8Array
      ? (leaf as Uint8Array)
      : serializer
      ? serializer(leaf as T)
      : new Uint8Array(32); // will never happen
    hash = await sha512(serialized, module);
    leavesHashed.set(hash, i * crypto_hash_sha512_BYTES);
    i++;
  }

  const ptr2 = module._malloc(crypto_hash_sha512_BYTES);
  const elementHash = new Uint8Array(
    module.HEAPU8.buffer,
    ptr2,
    crypto_hash_sha512_BYTES,
  );
  const elementSerialized = elementIsUint8Array
    ? element
    : serializer
    ? serializer(element)
    : new Uint8Array(32); // will never happen
  hash = await sha512(elementSerialized);
  elementHash.set(hash);

  const ptr3 = module._malloc(treeLen * (crypto_hash_sha512_BYTES + 1));
  const proof = new Uint8Array(
    module.HEAPU8.buffer,
    ptr3,
    treeLen * (crypto_hash_sha512_BYTES + 1),
  );

  const result = module._get_merkle_proof(
    treeLen,
    leavesHashed.byteOffset,
    elementHash.byteOffset,
    proof.byteOffset,
  );

  module._free(ptr1);
  module._free(ptr2);

  switch (result) {
    case -1: {
      module._free(ptr3);

      throw new Error("Element not in tree.");
    }

    case -2: {
      module._free(ptr3);

      throw new Error("Could not allocate memory for hashes helper array.");
    }

    case -3: {
      module._free(ptr3);

      throw new Error(
        "Could not allocate memory for hash concatenation helper array.",
      );
    }

    case -4: {
      module._free(ptr3);

      throw new Error("Could not calculate hash.");
    }

    default: {
      const proofArray = Uint8Array.from(proof.slice(0, result));
      module._free(ptr3);

      return proofArray;
    }
  }
};

export default getMerkleProof;

File: src/c/merkle/get_merkle_proof.c
// Copyright (C) 2022 Deliberative Technologies P.C.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#include "../utils/utils.h"

#include "../../../libsodium/src/libsodium/include/sodium/crypto_hash_sha512.h"

// The result is the proof length
__attribute__((used)) int
get_merkle_proof(
    const unsigned int LEAVES_LEN,
    const uint8_t leaves_hashed[LEAVES_LEN][crypto_hash_sha512_BYTES],
    const uint8_t element_hash[crypto_hash_sha512_BYTES],
    uint8_t proof[LEAVES_LEN][crypto_hash_sha512_BYTES + 1])
{
  size_t i, j, k;

  int32_t index[1];
  items_indexes_in_array(
      LEAVES_LEN, 1, leaves_hashed,
      (uint8_t(*)[crypto_hash_sha512_BYTES]) & element_hash[0], index);
  if (index[0] == -1) return -1;

  unsigned int element_of_interest = index[0];

  uint8_t(*hashes)[crypto_hash_sha512_BYTES]
      = malloc(sizeof(uint8_t[LEAVES_LEN][crypto_hash_sha512_BYTES]));
  if (hashes == NULL) return -2;

  memcpy(hashes, leaves_hashed, LEAVES_LEN * crypto_hash_sha512_BYTES);

  uint8_t(*concat_hashes)[crypto_hash_sha512_BYTES]
      = malloc(sizeof(uint8_t[2][crypto_hash_sha512_BYTES]));
  if (concat_hashes == NULL)
  {
    free(hashes);

    return -3;
  }

  unsigned int leaves = LEAVES_LEN;
  int res;
  bool odd_leaves;
  bool element_of_interest_found = false;

  // Counts the index of proof artifacts.
  k = 0;

  while (leaves > 0)
  {
    // Check if number of leaves is odd or even.
    odd_leaves = leaves % 2 != 0;

    // For every two leaves.
    for (i = 0, j = 0; i < leaves; i += 2, j++)
    {
      // If we are at the last position to the right of a tree with odd
      // number of leaves.
      if (odd_leaves && i + 1 == leaves)
      {
        memcpy(&concat_hashes[0], &hashes[i], crypto_hash_sha512_BYTES);
        // Concat leaf hash with itself.
        memcpy(&concat_hashes[1], &hashes[i], crypto_hash_sha512_BYTES);

        if (i == element_of_interest)
        {
          // Copy required hash of interest in the proof.
          memcpy(&proof[k], &hashes[i], crypto_hash_sha512_BYTES);
          // We do not care if left(0) or right(1) since hash of itself
          proof[k][crypto_hash_sha512_BYTES] = 0;

          k++;
          element_of_interest = j;
          element_of_interest_found = true;
        }
      }
      else
      {
        memcpy(&concat_hashes[0], &hashes[i], crypto_hash_sha512_BYTES);
        // In any other case concat leaf hash with the one on its right.
        memcpy(&concat_hashes[1], &hashes[i + 1], crypto_hash_sha512_BYTES);

        if (i == element_of_interest || i + 1 == element_of_interest)
        {
          if (i == element_of_interest)
          {
            memcpy(&proof[k], &hashes[i + 1], crypto_hash_sha512_BYTES);
            // Proof artifact needs to go to the right when concatenated with
            // element.
            proof[k][crypto_hash_sha512_BYTES] = 1;
          }
          else if (i + 1 == element_of_interest)
          {
            memcpy(&proof[k], &hashes[i], crypto_hash_sha512_BYTES);
            // Proof artifact needs to go to the left when concatenated with
            // element.
            proof[k][crypto_hash_sha512_BYTES] = 0;
          }

          k++;
          element_of_interest = j;
          element_of_interest_found = true;
        }
      }

      res = crypto_hash_sha512(hashes[j], *concat_hashes,
                               2 * crypto_hash_sha512_BYTES);
      if (res != 0)
      {
        free(hashes);
        free(concat_hashes);

        return -4;
      }
    }

    if (leaves == 1) break;

    leaves = ceil((double)leaves / 2);

    element_of_interest_found = false;
  }

  free(hashes);
  free(concat_hashes);

  return k * (crypto_hash_sha512_BYTES + 1);
}
